---
title: 'Исключения: Создание исключений из собственных функций | Документация Майкрософт'
ms.custom: ''
ms.date: 11/04/2016
ms.technology:
- cpp-mfc
ms.topic: conceptual
dev_langs:
- C++
helpviewer_keywords:
- throwing exceptions [MFC], from functions
- functions [MFC], throwing exceptions
- exceptions [MFC], throwing
ms.assetid: 492976e8-8804-4234-8e8f-30dffd0501be
author: mikeblome
ms.author: mblome
ms.workload:
- cplusplus
ms.openlocfilehash: a26bd11c2a37e3644333a95ed03d9182f7b32b87
ms.sourcegitcommit: a9dcbcc85b4c28eed280d8e451c494a00d8c4c25
ms.translationtype: MT
ms.contentlocale: ru-RU
ms.lasthandoff: 10/25/2018
ms.locfileid: "50066114"
---
# <a name="exceptions-throwing-exceptions-from-your-own-functions"></a>Исключения. Создание исключений из собственных функций

Это можно использовать в парадигме обработки исключений MFC исключительно для перехвата исключений, создаваемых с помощью функций в MFC или другие библиотеки. В дополнение к перехват исключения, создаваемые кодом библиотеки, может создавать исключения из собственного кода при написании функции, которые могут возникнуть исключительных условий.

При возникновении исключения выполнение текущей функции остановлена и переходит непосредственно к **catch** блок кадра корневое исключение. Механизм обработки исключений обход нормального завершения путь из функции. Таким образом необходимо убедиться удалить эти блоки памяти, которые будут удалены в нормального завершения.

#### <a name="to-throw-an-exception"></a>Для создания исключения

1. Используйте один из вспомогательных функций MFC, такие как `AfxThrowMemoryException`. Эти функции вызывают объект выделенные исключение соответствующего типа.

   В следующем примере функция пытается выделить два блока памяти и создает исключение, если либо выделение завершается с ошибкой:

   [!code-cpp[NVC_MFCExceptions#17](../mfc/codesnippet/cpp/exceptions-throwing-exceptions-from-your-own-functions_1.cpp)]

   В случае первого сбоя, вы можете просто исключение памяти. При первом выделении будет успешной, а вторая завершился неудачно, необходимо освободить первого блока распределения до создания исключения. Если удается выделения, вы можете продолжить обычно и освобождение блоков, при выходе из функции.

     - или

1. Используйте пользовательского исключения, чтобы указать условие проблемы. Элемент любого типа, даже целый класс, можно создавать как исключение.

   В следующем примере предпринимается попытка воспроизведения звука через звукового устройства и создает исключение, если возникает сбой.

   [!code-cpp[NVC_MFCExceptions#18](../mfc/codesnippet/cpp/exceptions-throwing-exceptions-from-your-own-functions_2.cpp)]

> [!NOTE]
>  MFC обработки исключений по умолчанию применяется только к указатели на `CException` объектов (и объектов `CException`-производные классы). Приведенный выше пример обходит механизм исключений MFC.

## <a name="see-also"></a>См. также

[Обработка исключений](../mfc/exception-handling-in-mfc.md)

